File: trading-dashboard/frontend/src/stores/market.js
import { writable } from 'svelte/store';

// Market sectors list
const SECTORS = [
	'Basic Materials',
	'Communication Services',
	'Consumer Cyclical',
	'Consumer Defensive',
	'Energy',
	'Financial Services',
	'Healthcare',
	'Industrials',
	'Real Estate',
	'Technology',
	'Utilities'
];

// Create the market store
function createMarketStore() {
	const { subscribe, set, update } = writable({
		overallRating: 0,
		sectorRatings: {},
		sectors: SECTORS,
		loading: false,
		lastSaved: null
	});

	return {
		subscribe,
		
		// Set overall market rating
		setOverallRating: (rating) => update(state => ({
			...state,
			overallRating: rating
		})),
		
		// Set individual sector rating
		setSectorRating: (sector, rating) => update(state => ({
			...state,
			sectorRatings: {
				...state.sectorRatings,
				[sector]: rating
			}
		})),
		
		// Load latest rating from backend
		loadLatestRating: async () => {
			update(state => ({ ...state, loading: true }));
			
			try {
				// Call the Wails backend API
				const rating = await window.go.main.App.GetLatestMarketRating();
				
				update(state => ({
					...state,
					overallRating: rating.overall_rating || 0,
					sectorRatings: rating.sector_ratings || {},
					lastSaved: rating.created_at,
					loading: false
				}));
			} catch (error) {
				console.error('Failed to load latest rating:', error);
				update(state => ({ ...state, loading: false }));
				throw error;
			}
		},
		
		// Save current ratings to backend
		saveRating: async () => {
			update(state => ({ ...state, loading: true }));
			
			try {
				let currentState;
				const unsubscribe = subscribe(state => currentState = state);
				unsubscribe();
				
				const request = {
					overall_rating: currentState.overallRating,
					sector_ratings: currentState.sectorRatings
				};
				
				const savedRating = await window.go.main.App.SaveMarketRating(request);
				
				update(state => ({
					...state,
					lastSaved: savedRating.created_at,
					loading: false
				}));
				
				return savedRating;
			} catch (error) {
				console.error('Failed to save rating:', error);
				update(state => ({ ...state, loading: false }));
				throw error;
			}
		},
		
		// Reset all ratings to zero
		resetRatings: () => update(state => ({
			...state,
			overallRating: 0,
			sectorRatings: {}
		})),
		
		// Get sectors list
		getSectors: () => SECTORS
	};
}

export const marketStore = createMarketStore(); 

File: trading-dashboard/backend/internal/models/market.go
package models

import (
	"time"
)

// MarketRating represents the overall market sentiment and associated sector ratings
type MarketRating struct {
	ID            int64              `json:"id"`
	OverallRating float64            `json:"overall_rating"`
	SectorRatings map[string]float64 `json:"sector_ratings"`
	CreatedAt     time.Time          `json:"created_at"`
	UpdatedAt     time.Time          `json:"updated_at"`
}

// SectorRating represents a single sector's sentiment rating
type SectorRating struct {
	ID             int64     `json:"id"`
	MarketRatingID int64     `json:"market_rating_id"`
	SectorName     string    `json:"sector_name"`
	Rating         float64   `json:"rating"`
	CreatedAt      time.Time `json:"created_at"`
}

// MarketRatingRequest represents the data structure for creating/updating market ratings
type MarketRatingRequest struct {
	OverallRating float64            `json:"overall_rating"`
	SectorRatings map[string]float64 `json:"sector_ratings"`
}

// ValidateRating checks if a rating is within the valid range (-3 to +3)
func ValidateRating(rating float64) bool {
	return rating >= -3.0 && rating <= 3.0
}

// GetSectorNames returns the standard list of market sectors
func GetSectorNames() []string {
	return []string{
		"Basic Materials",
		"Communication Services",
		"Consumer Cyclical",
		"Consumer Defensive",
		"Energy",
		"Financial Services",
		"Healthcare",
		"Industrials",
		"Real Estate",
		"Technology",
		"Utilities",
	}
}
